### PR TITLE
Support for GRPC

### LINKED ISSUE #278: gRPC implementation in server-node and router
# About Connext
At Connext, our goal is to build the cross-chain routing and micropayment layer of the decentralized web.
Vector sits on top of Ethereum, evm-compatible L2 blockchains, and other turing-complete chains, and enables instant, near free transfers that can be routed across chains and over liquidity in any asset.

# Task

The overall goal of this task is to create a gRPC-based communication layer on top of Vector `server-node`s. Currently, the server nodes have an HTTP REST interface for communication. This is okay for user-node actions, but in a high-throughput microservice architecture, we need something more performant and flexible.

The Vector `router` is a microservice that is designed to run in a stack with a server node. The router currently interacts with the server node through HTTP requests. We would like to enable bidirectional communication between the server node and the router, which will be easily enabled through gRPC.

# Requirements

- Create gRPC server in `server-node` that implements the API that is defined by the endpoints here: [https://github.com/connext/vector/blob/main/modules/server-node/src/index.ts](https://github.com/connext/vector/blob/main/modules/server-node/src/index.ts)
- Create a gRPC client in `router` that can communicate with the `server-node`.
- Create gRPC server in `router` which can push events back to the `server-node`.
- Make sure protobuf specs have corresponding Typescript types (even better if they are automatically generated!).
- Follow the code style of Vector repo.

# Bounty
A total of $200 DAI is allocated to this bounty.

# Links:
- More about connext & vector: https://github.com/connext/vector
- Chat at Discord: https://discord.com/channels/454734546869551114

### PR DESCRIPTION
**Pull request by janus**
## The Problem
Adding interfaces to you enable GRPC

closes #278

## The Solution
Generated interfaces for GRPC.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -15,6 +15,7 @@
   },
   "homepage": "https://github.com/connext/vector#readme",
   "devDependencies": {
+    "@protobuf-ts/plugin": "^1.0.12",
```
Comment (rhlsthrm) [inline package.json:18]: We shouldn't have these dependencies at the top level.

**Review by rhlsthrm — COMMENTED** [review]
Hi @janus, thanks for this! I'm curious how I can go about testing this. Can you write up some docs on it? For example we have example HTTP requests here: `modules/server-node/examples`. Would be great if we can have something similar for this.

File: package.json (hunk 1 above)
Comment (sanchaymittal) [reply package.json]: moved it to server-node dir package.json.

**Comment by rhlsthrm** [comment]
Hi @janus, I've invited you to be a collaborator on the repo. Can you make a branch directly off the main repo so that the team also can help push to the branch? That way we can give feedback and help make changes directly in the branch.

**Comment by janus** [comment]
> Hi @janus, I've invited you to be a collaborator on the repo. Can you make a branch directly off the main repo so that the team also can help push to the branch? That way we can give feedback and help make changes directly in the branch.

Sure, and thanks. I would do that once I get back to my laptop.

**Comment by janus** [comment]
> Hi @janus, thanks for this! I'm curious how I can go about testing this. Can you write up some docs on it? For example we have example HTTP requests here: `modules/server-node/examples`. Would be great if we can have something similar for this.

Okay....I would.

**Comment by rhlsthrm** [comment]
> > Hi @janus, I've invited you to be a collaborator on the repo. Can you make a branch directly off the main repo so that the team also can help push to the branch? That way we can give feedback and help make changes directly in the branch.
> 
> Sure, and thanks. I would do that once I get back to my laptop.

Never mind, this is not necessary. I am able to check out your fork. Just leave it as is.

https://stackoverflow.com/questions/5884784/how-to-pull-remote-branch-from-somebody-elses-repo

**Comment by sanchaymittal** [comment]
Hey @janus, How it's going?

**Comment by rhlsthrm** [comment]
Hi, @janus, are you working on this issue anymore? The progress has been slow and unresponsive recently.

**Comment by janus** [comment]
> Hi, @janus, are you working on this issue anymore? The progress has been slow and unresponsive recently.

Sorry , I had light fever.  I am getting back to work.

**Comment by sanchaymittal** [comment]
@janus Don't worry, We picked up the task on our end. We are merging this pr. Thanks for your hard work!

**Comment by rhlsthrm** [comment]
@janus can you share the exact command you used to generate the TS types from the proto files?

**Comment by janus** [comment]
> @janus can you share the exact command you used to generate the TS types from the proto files?

Sure... Once I get to my computer

**Comment by janus** [comment]
> @janus can you share the exact command you used to generate the TS types from the proto files?

`protoc  --plugin node_modules/.bin/protoc-gen-ts --ts_out modules/server-node/server/generated --ts_opt optimize_code_size  modules/server-node/server/proto/vector.proto `

**Comment by janus** [comment]
@sanchaymittal   Don't forget to follow this links, https://github.com/timostamm/protobuf-ts/blob/v1.0.0/MANUAL.md#googleprotobufany
and 
https://github.com/improbable-eng/ts-protoc-gen

**Comment by janus** [comment]
Thanks for your generosity.  I would find a way of paying you back.

**Comment by rhlsthrm** [comment]
Thanks for the links! That's exactly what we needed! I think I have things generally set up in our app now, so I will begin writing out tests and building it. Do you have experience with grpc-web? Would like to hear your thoughts.

**Comment by janus** [comment]
> Thanks for the links! That's exactly what we needed! I think I have things generally set up in our app now, so I will begin writing out tests and building it. Do you have experience with grpc-web? Would like to hear your thoughts.

Sorry for late reply. I have not done `grpc-web` before, but I don't think it is all that hard. And I was also thinking that it would reduce duplication if applied in this task.  Should I research it?

**Comment by rhlsthrm** [comment]
Hi @janus take a look at this: https://github.com/connext/vector/pull/326

I redid the protobuf and started building out the client and server. If you have a good idea of how to implement grpc-web into this we can chat about another bounty.

**Comment by janus** [comment]
Can we work together on this? Leave monetary part aside. I want to see this work.

**Pull request merged**